daemon/cmd/state.go

```diff
@@ -439,43 +444,44 @@ func (d *Daemon) allocateIPsLocked(ep *endpoint.Endpoint) (err error) {
 func (d *Daemon) initRestore(restoredEndpoints *endpointRestoreState) chan struct{} {
 	bootstrapStats.restore.Start()
 	var restoreComplete chan struct{}
+	var clusterSyncComplete chan struct{}
 	if option.Config.RestoreState {
 		// When we regenerate restored endpoints, it is guaranteed tha we have
 		// received the full list of policies present at the time the daemon
 		// is bootstrapped.
 		restoreComplete = d.regenerateRestoredEndpoints(restoredEndpoints)
-		go func() {
-			<-restoreComplete
-		}()
-
-		go func() {
-			if d.clientset.IsEnabled() {
-				// Also wait for all cluster mesh to be synchronized with the
-				// datapath before proceeding.
-				if d.clustermesh != nil {
-					err := d.clustermesh.ClustersSynced(d.ctx)
-					if err != nil {
-						log.WithError(err).Fatal("timeout while waiting for all clusters to be locally synchronized")
+		if d.clientset.IsEnabled() {
+			d.jobGroup.Add(
+				job.OneShot("setup-sync-lb-maps", func(ctx context.Context) error {
+					clusterSyncComplete = make(chan struct{})
+					// Also wait for all cluster mesh to be synchronized with the
+					// datapath before proceeding.
+					if d.clustermesh != nil {
+						err := d.clustermesh.ClustersSynced(d.ctx)
+						if err != nil {
+							log.WithError(err).Fatal("timeout while waiting for all clusters to be locally synchronized")
+						}
+						log.Debug("all clusters have been correctly synchronized locally")
 					}
-					log.Debug("all clusters have been correctly synchronized locally")
-				}
-				// Start controller which removes any leftover Kubernetes
+					close(clusterSyncComplete)
+					return nil
+				}),
+				// Start a job which removes any leftover Kubernetes
 				// services that may have been deleted while Cilium was not
-				// running. Once this controller succeeds, because it has no
-				// RunInterval specified, it will not run again unless updated
-				// elsewhere. This means that if, for instance, a user manually
+				// running. Since this is a OneShot job, it will not run again unless
+				// updated elsewhere. This means that if, for instance, a user manually
 				// adds a service via the CLI into the BPF maps, that it will
 				// not be cleaned up by the daemon until it restarts.
-				controller.NewManager().UpdateController("sync-lb-maps-with-k8s-services",
-					controller.ControllerParams{
-						DoFunc: func(ctx context.Context) error {
-							return d.svc.SyncWithK8sFinished(d.k8sWatcher.K8sSvcCache.EnsureService)
-						},
-						Context: d.ctx,
+				job.OneShot(
+					"sync-lb-maps-with-k8s-services",
+					func(ctx context.Context) error {
+						<-clusterSyncComplete
```

Comment:
I would not split the logic in two separate jobs, as they are strictly related. In particular, it is necessary to wait until all clusters are synchronized before triggering the service synchronization.

Reply:
OK will combine them.